File: frontend/src/lib/sdk/annotation-sdk.ts
/**
 * Annotation App JavaScript SDK
 * Provides a TypeScript-friendly interface for interacting with the Annotation API
 */

export interface APIConfig {
  apiKey: string;
  baseURL?: string;
  timeout?: number;
  retries?: number;
}

export interface APIResponse<T = unknown> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

export interface Document {
  id: string;
  title: string;
  content: string;
  created_at: string;
  updated_at: string;
  annotation_count: number;
  status: "pending" | "in_progress" | "completed";
  metadata?: Record<string, unknown>;
}

export interface Annotation {
  id: string;
  document_id: string;
  start: number;
  end: number;
  text: string;
  label: string;
  confidence?: number;
  annotator_id?: string;
  created_at: string;
  metadata?: Record<string, unknown>;
}

export interface ExportOptions {
  format: "json" | "conll" | "csv" | "standoff";
  include_metadata?: boolean;
  include_statistics?: boolean;
  filter_labels?: string[];
  encoding_scheme?: "BIO" | "BILOU" | "IO";
  csv_separator?: string;
}

export interface ExportJob {
  id: string;
  status: "pending" | "processing" | "completed" | "failed";
  document_id: string;
  format: string;
  created_at: string;
  completed_at?: string;
  download_url?: string;
  error_message?: string;
  progress?: number;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  pages: number;
  limit: number;
}

export interface ListOptions {
  page?: number;
  limit?: number;
  search?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
  [key: string]: unknown;
}

/**
 * HTTP Client for making API requests
 */
class HTTPClient {
  private config: Required<APIConfig>;

  constructor(config: APIConfig) {
    this.config = {
      baseURL: "https://api.annotation-app.com",
      timeout: 30000,
      retries: 3,
      ...config,
    };
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<APIResponse<T>> {
    const url = `${this.config.baseURL}${endpoint}`;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url, {
        ...options,
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.config.apiKey}`,
          ...options.headers,
        },
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      const data = await response.json();

      if (!response.ok) {
        return {
          success: false,
          error:
            data.error || `HTTP ${response.status}: ${response.statusText}`,
        };
      }

      return {
        success: true,
        data,
      };
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof Error) {
        return {
          success: false,
          error: error.message,
        };
      }

      return {
        success: false,
        error: "Unknown error occurred",
      };
    }
  }

  async get<T>(
    endpoint: string,
    params?: Record<string, unknown>
  ): Promise<APIResponse<T>> {
    const url = new URL(endpoint, this.config.baseURL);
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    return this.request<T>(url.pathname + url.search);
  }

  async post<T>(endpoint: string, data?: unknown): Promise<APIResponse<T>> {
    return this.request<T>(endpoint, {
      method: "POST",
      body: data ? JSON.stringify(data) : undefined,
    });
  }

  async put<T>(endpoint: string, data?: unknown): Promise<APIResponse<T>> {
    return this.request<T>(endpoint, {
      method: "PUT",
      body: data ? JSON.stringify(data) : undefined,
    });
  }

  async delete<T>(endpoint: string): Promise<APIResponse<T>> {
    return this.request<T>(endpoint, {
      method: "DELETE",
    });
  }
}

/**
 * Documents API client
 */
export class DocumentsAPI {
  constructor(private client: HTTPClient) {}

  /**
   * List all documents
   */
  async list(
    options: ListOptions = {}
  ): Promise<APIResponse<PaginatedResponse<Document>>> {
    return this.client.get<PaginatedResponse<Document>>(
      "/api/documents",
      options
    );
  }

  /**
   * Get a specific document by ID
   */
  async get(id: string): Promise<APIResponse<Document>> {
    return this.client.get<Document>(`/api/documents/${id}`);
  }

  /**
   * Create a new document
   */
  async create(document: Partial<Document>): Promise<APIResponse<Document>> {
    return this.client.post<Document>("/api/documents", document);
  }

  /**
   * Update an existing document
   */
  async update(
    id: string,
    updates: Partial<Document>
  ): Promise<APIResponse<Document>> {
    return this.client.put<Document>(`/api/documents/${id}`, updates);
  }

  /**
   * Delete a document
   */
  async delete(id: string): Promise<APIResponse<void>> {
    return this.client.delete<void>(`/api/documents/${id}`);
  }

  /**
   * Export document annotations
   */
  async export(
    id: string,
    options: ExportOptions
  ): Promise<APIResponse<ExportJob>> {
    return this.client.post<ExportJob>(`/api/documents/${id}/export`, options);
  }

  /**
   * Get export job status
   */
  async getExportJob(jobId: string): Promise<APIResponse<ExportJob>> {
    return this.client.get<ExportJob>(`/api/export-jobs/${jobId}`);
  }

  /**
   * Download export file
   */
  async downloadExport(jobId: string): Promise<APIResponse<Blob>> {
    // Special handling for file downloads
    const url = `${this.client["config"].baseURL}/api/export-jobs/${jobId}/download`;

    try {
      const response = await fetch(url, {
        headers: {
          Authorization: `Bearer ${this.client["config"].apiKey}`,
        },
      });

      if (!response.ok) {
        return {
          success: false,
          error: `HTTP ${response.status}: ${response.statusText}`,
        };
      }

      const blob = await response.blob();
      return {
        success: true,
        data: blob,
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : "Download failed",
      };
    }
  }
}

/**
 * Annotations API client
 */
export class AnnotationsAPI {
  constructor(private client: HTTPClient) {}

  /**
   * List annotations for a document
   */
  async list(
    documentId: string,
    options: ListOptions & { label?: string } = {}
  ): Promise<APIResponse<PaginatedResponse<Annotation>>> {
    return this.client.get<PaginatedResponse<Annotation>>(
      `/api/documents/${documentId}/annotations`,
      options
    );
  }

  /**
   * Get a specific annotation
   */
  async get(id: string): Promise<APIResponse<Annotation>> {
    return this.client.get<Annotation>(`/api/annotations/${id}`);
  }

  /**
   * Create a new annotation
   */
  async create(
    annotation: Partial<Annotation>
  ): Promise<APIResponse<Annotation>> {
    return this.client.post<Annotation>("/api/annotations", annotation);
  }

  /**
   * Update an existing annotation
   */
  async update(
    id: string,
    updates: Partial<Annotation>
  ): Promise<APIResponse<Annotation>> {
    return this.client.put<Annotation>(`/api/annotations/${id}`, updates);
  }

  /**
   * Delete an annotation
   */
  async delete(id: string): Promise<APIResponse<void>> {
    return this.client.delete<void>(`/api/annotations/${id}`);
  }

  /**
   * Bulk create annotations
   */
  async bulkCreate(
    annotations: Partial<Annotation>[]
  ): Promise<APIResponse<Annotation[]>> {
    return this.client.post<Annotation[]>("/api/annotations/bulk", {
      annotations,
    });
  }
}

/**
 * Main SDK client
 */
export class AnnotationAPI {
  private client: HTTPClient;
  public documents: DocumentsAPI;
  public annotations: AnnotationsAPI;

  constructor(config: APIConfig) {
    this.client = new HTTPClient(config);
    this.documents = new DocumentsAPI(this.client);
    this.annotations = new AnnotationsAPI(this.client);
  }

  /**
   * Test API connection
   */
  async ping(): Promise<APIResponse<{ message: string; timestamp: string }>> {
    return this.client.get("/api/ping");
  }

  /**
   * Get API status and version
   */
  async status(): Promise<APIResponse<{ version: string; status: string }>> {
    return this.client.get("/api/status");
  }
}

/**
 * Utility functions
 */
export class AnnotationUtils {
  /**
   * Calculate annotation statistics
   */
  static calculateStats(annotations: Annotation[]) {
    const labelCounts: Record<string, number> = {};
    const annotatorCounts: Record<string, number> = {};
    let totalLength = 0;

    for (const annotation of annotations) {
      // Count labels
      labelCounts[annotation.label] = (labelCounts[annotation.label] || 0) + 1;

      // Count annotators
      if (annotation.annotator_id) {
        annotatorCounts[annotation.annotator_id] =
          (annotatorCounts[annotation.annotator_id] || 0) + 1;
      }

      // Calculate text length
      totalLength += annotation.text.length;
    }

    return {
      total: annotations.length,
      labels: labelCounts,
      annotators: annotatorCounts,
      averageLength:
        annotations.length > 0 ? totalLength / annotations.length : 0,
      totalLength,
    };
  }

  /**
   * Validate annotation overlap
   */
  static findOverlaps(annotations: Annotation[]): Array<{
    annotation1: Annotation;
    annotation2: Annotation;
    overlapStart: number;
    overlapEnd: number;
  }> {
    const overlaps = [];

    for (let i = 0; i < annotations.length; i++) {
      for (let j = i + 1; j < annotations.length; j++) {
        const ann1 = annotations[i];
        const ann2 = annotations[j];

        // Check if they're for the same document
        if (ann1.document_id !== ann2.document_id) continue;

        // Check for overlap
        const overlapStart = Math.max(ann1.start, ann2.start);
        const overlapEnd = Math.min(ann1.end, ann2.end);

        if (overlapStart < overlapEnd) {
          overlaps.push({
            annotation1: ann1,
            annotation2: ann2,
            overlapStart,
            overlapEnd,
          });
        }
      }
    }

    return overlaps;
  }

  /**
   * Convert annotations to CoNLL format
   */
  static toCoNLL(
    text: string,
    annotations: Annotation[],
    scheme: "BIO" | "BILOU" | "IO" = "BIO"
  ): string {
    const tokens = text.split(/\s+/);
    const tokenAnnotations: string[] = new Array(tokens.length).fill("O");

    let currentPos = 0;
    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];
      const tokenStart = text.indexOf(token, currentPos);
      const tokenEnd = tokenStart + token.length;

      // Find annotations that cover this token
      for (const annotation of annotations) {
        if (annotation.start <= tokenStart && annotation.end >= tokenEnd) {
          const isFirst =
            i === 0 || !tokenAnnotations[i - 1].endsWith(annotation.label);

          if (scheme === "BIO") {
            tokenAnnotations[i] = `${isFirst ? "B" : "I"}-${annotation.label}`;
          } else if (scheme === "IO") {
            tokenAnnotations[i] = `I-${annotation.label}`;
          } else if (scheme === "BILOU") {
            const isLast =
              i === tokens.length - 1 ||
              !annotations.some(
                (a) =>
                  a.label === annotation.label &&
                  a.start <= text.indexOf(tokens[i + 1], tokenEnd) &&
                  a.end >=
                    text.indexOf(tokens[i + 1], tokenEnd) + tokens[i + 1].length
              );

            if (isFirst && isLast) {
              tokenAnnotations[i] = `U-${annotation.label}`;
            } else if (isFirst) {
              tokenAnnotations[i] = `B-${annotation.label}`;
            } else if (isLast) {
              tokenAnnotations[i] = `L-${annotation.label}`;
            } else {
              tokenAnnotations[i] = `I-${annotation.label}`;
            }
          }
          break; // Use first matching annotation
        }
      }

      currentPos = tokenEnd;
    }

    // Build CoNLL output
    const lines = [];
    for (let i = 0; i < tokens.length; i++) {
      lines.push(`${tokens[i]}\t${tokenAnnotations[i]}`);
    }

    return lines.join("\n");
  }
}

// Export default instance for convenience
export default AnnotationAPI;

File: frontend/src/lib/sdk/annotation_sdk.py
"""
Annotation App Python SDK
Provides a Python interface for interacting with the Annotation API
"""

import json
import time
from typing import Dict, List, Optional, Union, Any
from dataclasses import dataclass, asdict
from urllib.parse import urljoin, urlencode
import requests


@dataclass
class APIConfig:
    """Configuration for the API client"""
    api_key: str
    base_url: str = "https://api.annotation-app.com"
    timeout: int = 30
    retries: int = 3


@dataclass
class APIResponse:
    """Response wrapper for API calls"""
    success: bool
    data: Optional[Any] = None
    error: Optional[str] = None
    message: Optional[str] = None


@dataclass
class Document:
    """Document model"""
    id: str
    title: str
    content: str
    created_at: str
    updated_at: str
    annotation_count: int
    status: str
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class Annotation:
    """Annotation model"""
    id: str
    document_id: str
    start: int
    end: int
    text: str
    label: str
    confidence: Optional[float] = None
    annotator_id: Optional[str] = None
    created_at: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class ExportOptions:
    """Export configuration options"""
    format: str  # json, conll, csv, standoff
    include_metadata: bool = True
    include_statistics: bool = True
    filter_labels: Optional[List[str]] = None
    encoding_scheme: str = "BIO"  # BIO, BILOU, IO
    csv_separator: str = ","


@dataclass
class ExportJob:
    """Export job status"""
    id: str
    status: str  # pending, processing, completed, failed
    document_id: str
    format: str
    created_at: str
    completed_at: Optional[str] = None
    download_url: Optional[str] = None
    error_message: Optional[str] = None
    progress: Optional[int] = None


@dataclass
class PaginatedResponse:
    """Paginated response wrapper"""
    items: List[Any]
    total: int
    page: int
    pages: int
    limit: int


class HTTPClient:
    """HTTP client for making API requests"""
    
    def __init__(self, config: APIConfig):
        self.config = config
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {config.api_key}",
            "Content-Type": "application/json",
        })
    
    def _request(self, method: str, endpoint: str, **kwargs) -> APIResponse:
        """Make an HTTP request with error handling and retries"""
        url = urljoin(self.config.base_url, endpoint)
        
        for attempt in range(self.config.retries):
            try:
                response = self.session.request(
                    method=method,
                    url=url,
                    timeout=self.config.timeout,
                    **kwargs
                )
                
                # Handle JSON response
                try:
                    data = response.json()
                except json.JSONDecodeError:
                    data = response.text
                
                if response.ok:
                    return APIResponse(success=True, data=data)
                else:
                    error_msg = data.get("error", f"HTTP {response.status_code}: {response.reason}") if isinstance(data, dict) else str(data)
                    return APIResponse(success=False, error=error_msg)
                    
            except requests.exceptions.RequestException as e:
                if attempt == self.config.retries - 1:  # Last attempt
                    return APIResponse(success=False, error=str(e))
                time.sleep(2 ** attempt)  # Exponential backoff
        
        return APIResponse(success=False, error="Max retries exceeded")
    
    def get(self, endpoint: str, params: Optional[Dict[str, Any]] = None) -> APIResponse:
        """Make a GET request"""
        return self._request("GET", endpoint, params=params)
    
    def post(self, endpoint: str, data: Optional[Dict[str, Any]] = None) -> APIResponse:
        """Make a POST request"""
        json_data = data if data is not None else {}
        return self._request("POST", endpoint, json=json_data)
    
    def put(self, endpoint: str, data: Optional[Dict[str, Any]] = None) -> APIResponse:
        """Make a PUT request"""
        json_data = data if data is not None else {}
        return self._request("PUT", endpoint, json=json_data)
    
    def delete(self, endpoint: str) -> APIResponse:
        """Make a DELETE request"""
        return self._request("DELETE", endpoint)


class DocumentsAPI:
    """Documents API client"""
    
    def __init__(self, client: HTTPClient):
        self.client = client
    
    def list(self, page: int = 1, limit: int = 20, search: Optional[str] = None, 
             sort_by: Optional[str] = None, sort_order: str = "desc") -> APIResponse:
        """List all documents"""
        params = {
            "page": page,
            "limit": limit,
            "sort_order": sort_order,
        }
        if search:
            params["search"] = search
        if sort_by:
            params["sort_by"] = sort_by
            
        return self.client.get("/api/documents", params)
    
    def get(self, document_id: str) -> APIResponse:
        """Get a specific document by ID"""
        return self.client.get(f"/api/documents/{document_id}")
    
    def create(self, document: Dict[str, Any]) -> APIResponse:
        """Create a new document"""
        return self.client.post("/api/documents", document)
    
    def update(self, document_id: str, updates: Dict[str, Any]) -> APIResponse:
        """Update an existing document"""
        return self.client.put(f"/api/documents/{document_id}", updates)
    
    def delete(self, document_id: str) -> APIResponse:
        """Delete a document"""
        return self.client.delete(f"/api/documents/{document_id}")
    
    def export(self, document_id: str, options: ExportOptions) -> APIResponse:
        """Export document annotations"""
        export_data = asdict(options)
        return self.client.post(f"/api/documents/{document_id}/export", export_data)
    
    def get_export_job(self, job_id: str) -> APIResponse:
        """Get export job status"""
        return self.client.get(f"/api/export-jobs/{job_id}")
    
    def download_export(self, job_id: str) -> APIResponse:
        """Download export file"""
        try:
            url = urljoin(self.client.config.base_url, f"/api/export-jobs/{job_id}/download")
            response = self.client.session.get(url, stream=True)
            
            if response.ok:
                return APIResponse(success=True, data=response.content)
            else:
                return APIResponse(success=False, error=f"HTTP {response.status_code}: {response.reason}")
                
        except requests.exceptions.RequestException as e:
            return APIResponse(success=False, error=str(e))


class AnnotationsAPI:
    """Annotations API client"""
    
    def __init__(self, client: HTTPClient):
        self.client = client
    
    def list(self, document_id: str, page: int = 1, limit: int = 100, 
             label: Optional[str] = None) -> APIResponse:
        """List annotations for a document"""
        params = {
            "page": page,
            "limit": limit,
        }
        if label:
            params["label"] = label
            
        return self.client.get(f"/api/documents/{document_id}/annotations", params)
    
    def get(self, annotation_id: str) -> APIResponse:
        """Get a specific annotation"""
        return self.client.get(f"/api/annotations/{annotation_id}")
    
    def create(self, annotation: Dict[str, Any]) -> APIResponse:
        """Create a new annotation"""
        return self.client.post("/api/annotations", annotation)
    
    def update(self, annotation_id: str, updates: Dict[str, Any]) -> APIResponse:
        """Update an existing annotation"""
        return self.client.put(f"/api/annotations/{annotation_id}", updates)
    
    def delete(self, annotation_id: str) -> APIResponse:
        """Delete an annotation"""
        return self.client.delete(f"/api/annotations/{annotation_id}")
    
    def bulk_create(self, annotations: List[Dict[str, Any]]) -> APIResponse:
        """Bulk create annotations"""
        return self.client.post("/api/annotations/bulk", {"annotations": annotations})


class AnnotationAPI:
    """Main SDK client"""
    
    def __init__(self, api_key: str, base_url: str = "https://api.annotation-app.com"):
        """Initialize the API client"""
        config = APIConfig(api_key=api_key, base_url=base_url)
        self.client = HTTPClient(config)
        self.documents = DocumentsAPI(self.client)
        self.annotations = AnnotationsAPI(self.client)
    
    def ping(self) -> APIResponse:
        """Test API connection"""
        return self.client.get("/api/ping")
    
    def status(self) -> APIResponse:
        """Get API status and version"""
        return self.client.get("/api/status")


class AnnotationUtils:
    """Utility functions for working with annotations"""
    
    @staticmethod
    def calculate_stats(annotations: List[Annotation]) -> Dict[str, Any]:
        """Calculate annotation statistics"""
        if not annotations:
            return {
                "total": 0,
                "labels": {},
                "annotators": {},
                "average_length": 0,
                "total_length": 0,
            }
        
        label_counts = {}
        annotator_counts = {}
        total_length = 0
        
        for annotation in annotations:
            # Count labels
            label_counts[annotation.label] = label_counts.get(annotation.label, 0) + 1
            
            # Count annotators
            if annotation.annotator_id:
                annotator_counts[annotation.annotator_id] = annotator_counts.get(annotation.annotator_id, 0) + 1
            
            # Calculate text length
            total_length += len(annotation.text)
        
        return {
            "total": len(annotations),
            "labels": label_counts,
            "annotators": annotator_counts,
            "average_length": total_length / len(annotations),
            "total_length": total_length,
        }
    
    @staticmethod
    def find_overlaps(annotations: List[Annotation]) -> List[Dict[str, Any]]:
        """Find overlapping annotations"""
        overlaps = []
        
        for i, ann1 in enumerate(annotations):
            for ann2 in annotations[i + 1:]:
                # Check if they're for the same document
                if ann1.document_id != ann2.document_id:
                    continue
                
                # Check for overlap
                overlap_start = max(ann1.start, ann2.start)
                overlap_end = min(ann1.end, ann2.end)
                
                if overlap_start < overlap_end:
                    overlaps.append({
                        "annotation1": ann1,
                        "annotation2": ann2,
                        "overlap_start": overlap_start,
                        "overlap_end": overlap_end,
                    })
        
        return overlaps
    
    @staticmethod
    def to_conll(text: str, annotations: List[Annotation], scheme: str = "BIO") -> str:
        """Convert annotations to CoNLL format"""
        tokens = text.split()
        token_annotations = ["O"] * len(tokens)
        
        current_pos = 0
        for i, token in enumerate(tokens):
            token_start = text.find(token, current_pos)
            token_end = token_start + len(token)
            
            # Find annotations that cover this token
            for annotation in annotations:
                if annotation.start <= token_start and annotation.end >= token_end:
                    is_first = i == 0 or not token_annotations[i - 1].endswith(annotation.label)
                    
                    if scheme == "BIO":
                        token_annotations[i] = f"{'B' if is_first else 'I'}-{annotation.label}"
                    elif scheme == "IO":
                        token_annotations[i] = f"I-{annotation.label}"
                    elif scheme == "BILOU":
                        # Find if this is the last token of the annotation
                        is_last = i == len(tokens) - 1 or not any(
                            a.label == annotation.label and 
                            a.start <= text.find(tokens[i + 1], token_end) and 
                            a.end >= text.find(tokens[i + 1], token_end) + len(tokens[i + 1])
                            for a in annotations
                        )
                        
                        if is_first and is_last:
                            token_annotations[i] = f"U-{annotation.label}"
                        elif is_first:
                            token_annotations[i] = f"B-{annotation.label}"
                        elif is_last:
                            token_annotations[i] = f"L-{annotation.label}"
                        else:
                            token_annotations[i] = f"I-{annotation.label}"
                    break  # Use first matching annotation
            
            current_pos = token_end
        
        # Build CoNLL output
        lines = []
        for token, annotation in zip(tokens, token_annotations):
            lines.append(f"{token}\t{annotation}")
        
        return "\n".join(lines)
    
    @staticmethod
    def from_conll(conll_text: str) -> List[Dict[str, Any]]:
        """Parse CoNLL format to annotations"""
        annotations = []
        lines = conll_text.strip().split("\n")
        tokens = []
        labels = []
        
        for line in lines:
            if line.strip():
                parts = line.split("\t")
                if len(parts) >= 2:
                    tokens.append(parts[0])
                    labels.append(parts[1])
        
        # Convert token-level labels to span annotations
        current_annotation = None
        text_pos = 0
        
        for i, (token, label) in enumerate(zip(tokens, labels)):
            if label != "O":
                prefix, label_name = label.split("-", 1) if "-" in label else ("I", label)
                
                if prefix in ["B", "U"] or current_annotation is None or current_annotation["label"] != label_name:
                    # Start new annotation
                    if current_annotation:
                        annotations.append(current_annotation)
                    
                    current_annotation = {
                        "start": text_pos,
                        "end": text_pos + len(token),
                        "text": token,
                        "label": label_name,
                    }
                else:
                    # Continue current annotation
                    current_annotation["end"] = text_pos + len(token)
                    current_annotation["text"] += " " + token
                
                if prefix in ["L", "U"]:
                    # End current annotation
                    annotations.append(current_annotation)
                    current_annotation = None
            else:
                # No annotation for this token
                if current_annotation:
                    annotations.append(current_annotation)
                    current_annotation = None
            
            text_pos += len(token) + 1  # +1 for space
        
        # Add final annotation if exists
        if current_annotation:
            annotations.append(current_annotation)
        
        return annotations


# Example usage
if __name__ == "__main__":
    # Initialize the client
    client = AnnotationAPI(api_key="your_api_key_here")
    
    # Test connection
    ping_result = client.ping()
    print(f"Ping result: {ping_result}")
    
    # List documents
    docs_result = client.documents.list(limit=10)
    if docs_result.success:
        print(f"Found {len(docs_result.data['items'])} documents")
    else:
        print(f"Error: {docs_result.error}")
    
    # Create a document
    new_doc = {
        "title": "Test Document",
        "content": "This is a test document for annotation.",
        "metadata": {"source": "api_test"}
    }
    
    create_result = client.documents.create(new_doc)
    if create_result.success:
        doc_id = create_result.data["id"]
        print(f"Created document with ID: {doc_id}")
        
        # Create an annotation
        annotation = {
            "document_id": doc_id,
            "start": 10,
            "end": 14,
            "text": "test",
            "label": "KEYWORD"
        }
        
        ann_result = client.annotations.create(annotation)
        if ann_result.success:
            print(f"Created annotation: {ann_result.data}")
        else:
            print(f"Error creating annotation: {ann_result.error}")
    else:
        print(f"Error creating document: {create_result.error}")
